### DISCUSSION TITLE
How to extend Canvas for dynamic Workflows?

### DISCUSSION THREAD

**Discussion by kavink** [celery/celery#7108] [Ideas]
Celery has very simple primitives for workflows, I want basically create dynamic workflows from config(which change very often). I did find [Selinon](https://github.com/selinon/selinon) which looks very close to what I want.

Wondering if there are any future plans for Celery to make the Workflows more extensible or should I try to integrate Selinon or What would you recommend if I need to basically dynamically create a DAG of dynamic tasks?

**Comment by WoaDmulL** [comment]
You can create dynamic workflows using [this](http://docs.celeryproject.org/en/master/reference/celery.app.task.html#celery.app.task.Task.replace).

**Comment by kavink** [comment]
@WoaDmulL Where can I find example/documentation for `replace` ? Still trying to understand what it does and how it works, So I can leverage it. Thanks a lot !!

**Comment by WoaDmulL** [comment]
There's no documentation, yet...except of what I gave you before.
The main feature is that you can replace a task with another signature/task.

So for example, you can build a chain of another tasks and replace the previous task with it.

```
Task A:
    signature = chain(B, C, D, ...)
    raise self.replace(signature)

Task B:
    // you can replace it again with more subtasks or just process something that will return a value to be passed to the next task in the main chain

Task C:
    same as B

Task D:
    same as B

.
.
.
```
The execution order will be Task A, B, C, D, ... . Everything will run in order ( asynchronously ).

You can generate groups, chains or chords dynamically and replace them ( like an extension tree ).

**Comment by kavink** [comment]
@WoaDmulL Awesome thanks! Basically will end up creating a DAG, how do I get each of the tasks result or output or status ? Also im assuming if a task fails, the workflow will stop at failed and won't trigger tasks after failed ones right ?

**Comment by WoaDmulL** [comment]
For each task is assigned an UUID and each subtask has assigned a parent_id and a root_id. The root_id is always the task A ( from the example ), and the parent_id the id inherited from the previous task ( Task B parent_id is Task A id ). You can store them and use the signals or the on_success/on_failure methods to trace tasks' states. You could also use link_error option from the signature method  and on_error method to do something else when a task fails. 
If a subtask fails and its exception is not catched, then the workflow also fails, so the tasks matching the parent_id and root_id should be marked as failed too. This is kinda complex.

Another versions is to be sure that the workflow completely runs and each task finishes with success.

**Comment by jenstroeger** [comment]
@WoaDmulL and @kavink, you may keep in mind https://github.com/celery/celery/issues/4299 and https://github.com/celery/celery/issues/4642 when working with canvas elements.

**Comment by WoaDmulL** [comment]
@jenstroeger **Also** both link_error/on_error are not working with class based tasks ( #4022 ). **Also** chords will fail in a workflow if the worker stops and you have to catch both WorkerLostError and WorkerShutdown and to retry the tasks. **Also** the way you create signatures has to be completely lazy, so a `chord(header)(body)` won't work cause it will be executed after its generation and not at replace, but a `chord((header), body)` will do the trick. And many more others.

EDIT: You can't `run()` a workflow that uses `self.replace()`.
